Proposal: Replacements and Patch value in the structured data (#4558)
kubernetes-sigs/kustomize: proposals/22-03-value-in-the-structured-data.md

Use the status field to record the results of the latest review:
- implementable: The default for this repo. If the proposal is merged, you can start working on it.
- deferred: The proposal may be accepted in the future, but it has been shelved for the time being.
A new PR must be opened to update the proposal and gain reviewer consensus before work can begin.
- withdrawn: The author changed their mind and no longer wants to pursue the proposal.
A new PR must be opened to update the proposal and gain reviewer consensus before work can begin.
- rejected: This proposal should not be implemented.
- replaced: If you submit a new proposal that supersedes an older one,
update the older one's status to "replaced by <link>".
-->

## Summary

This proposal decides the interfaces to change values in the structured data (like json,yaml) inside a Kubernetes objects' field value and implements changing function target a few formats (mainly json).

## Motivation

<!--
If this proposal is an expansion of an existing GitHub issue, link to it here.
-->

Fields in Kubernetes objects sometimes include values formatted by structured data like json and yaml.

> kustomize can strong patch with Kubernetes objects, but kustomize can't manipulate one value on structured, formatted data in the Kubernetes object's yaml field.

Author: I'll write to `This is an expected behavior, but kustomize will be very helpful if it can change`.
Is it an eligible answer to your review?

> kustomize can strong patch with Kubernetes objects, but kustomize can't manipulate one value on structured, formatted data in the Kubernetes object's yaml field.

Reviewer: To be very clear, I would put here that we are not changing the fact that kustomize _does_ support structured edits, and it specifically does not support unstructured edits. We are proposing a small well-scoped exception to allow edits to structured JSON within ConfigMap strings.

> kustomize can strong patch with Kubernetes objects, but kustomize can't manipulate one value on structured, formatted data in the Kubernetes object's yaml field.

Author: I've updated this sentence. Could you recheck it?

Example, kustomize can't change value `"REPLACE_TARGET_HOSTNAME"` in this yaml file.
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: target-configmap
data:
  config.json: |-
    {"config": {
      "id": "42",
      "hostname": "REPLACE_TARGET_HOSTNAME"
    }}
```

This function has been wanted for a long time.
- https://github.com/kubernetes-sigs/kustomize/issues/680
- https://github.com/kubernetes-sigs/kustomize/issues/3787
- https://github.com/kubernetes-sigs/kustomize/issues/4517

This function can be an alternative deprecated [vars](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/vars/) function in many use cases.

**Goals:**
<!--
List the specific goals of the proposal. What is it trying to achieve? How will we
know that this has succeeded?
-->

1. Provide the way to update values in the structured data like kubernetes objects with many formats.
1. Be able to add replacement able format after.

**Non-goals:**
<!--
What is out of scope for this proposal? Listing non-goals helps to focus discussion
and make progress.
-->

> 1. Do not provide to perfectly alternative for deprecated [vars](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/vars/).

Author: I'll change it to https://github.com/kubernetes-sigs/kustomize/pull/4558/commits/83fce9e6b5f28f7f3df52b488dc39f7da2cb9546#diff-2b5a81ecf356afb37e2653ac1e2daf0b6af6cd126aec835df3c25b2a8be29389R89.
Sorry, I couldn't find enough items to write this section when I wrote this document.

e.g.:
- Does this proposal require new kinds, fields or CLI flags?
- Will this feature require extending the public interface of Kustomize's Go packages?
(it's ok if you're not sure yet)

A proof of concept PR is NOT required but is preferable to including large amounts of code
inline here, if you feel such implementation details are required to adequately explain your design.
If you have a PR, link to it at the top of this section.
-->

I propose to add options for replacing the value in structured data to replacements function. My sample implementation is [here](https://github.com/kubernetes-sigs/kustomize/pull/4518).
This idea is add two parameter `format` and `formatPath` to [options](https://github.com/kubernetes-sigs/kustomize/blob/8668691ade05bc17b3c6f44bcd4723735033196e/api/types/replacement.go#L67-L80) in replacement [TargetSelector](https://github.com/kubernetes-sigs/kustomize/blob/8668691ade05bc17b3c6f44bcd4723735033196e/api/types/replacement.go#L52-L64). The `format` option is used by select to structured data format like "json" or yaml, and The `formatPath` option is "path" to target to change values in structured data with selected format from `format` option.

> Example.

Reviewer: The example can go under user stories. I suggest one user story for `replacements` and one user story for `configMapGenerator`.

Author: I'll be moving.

```yaml
## source
apiVersion: v1
kind: ConfigMap
metadata:
  name: source-configmap
data:
  HOSTNAME: www.example.com
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: target-configmap
data:
  config.json: |-
    {"config": {
      "id": "42",
      "hostname": "REPLACE_TARGET_HOSTNAME"
    }}
```

```yaml
## replacement
replacements:
- source:
    kind: ConfigMap
    name: source-configmap
    fieldPath: data.HOSTNAME
  targets:
  - select:
      kind: ConfigMap
      name: target-configmap
    fieldPaths:
    - data.config\.json
    options:
      format: 'json'

>       formatPath: '/config/hostname'

Reviewer: IMO this syntax LGTM but I may need to think about it more when comparing to the ConfigMapGenerator syntax. 

I also think this needs some more details. For example, are we planning to extend `format` in the future to support other formats, like YAML? And maybe some details about what happens if one of these fields is present without the other, are there default values or do we throw an error?

Author: I tried to add details.

  At minimum, you should fill in the "Summary" and "Motivation" sections.
- [ ] **Create a PR.**
  Ping `@kubernetes-sigs/kustomize-admins` and `@kubernetes-sigs/kustomize-maintainers`.
-->

# Replacements and Patch value in the structured data.

**Authors**:
- koba1t

**Reviewers**:
- natasha41575
- knverey

**Status**: implementable
<!--
In general, all proposals made should be merged for the record, whether or not they are accepted.
Use the status field to record the results of the latest review:
- implementable: The default for this repo. If the proposal is merged, you can start working on it.
- deferred: The proposal may be accepted in the future, but it has been shelved for the time being.
A new PR must be opened to update the proposal and gain reviewer consensus before work can begin.
- withdrawn: The author changed their mind and no longer wants to pursue the proposal.
A new PR must be opened to update the proposal and gain reviewer consensus before work can begin.
- rejected: This proposal should not be implemented.
- replaced: If you submit a new proposal that supersedes an older one,
update the older one's status to "replaced by <link>".
-->

## Summary

> This proposal decides the interfaces to change values in the structured data (like json,yaml) inside a Kubernetes objects' field value and implements changing function target a few formats (mainly json).

Reviewer: ```suggestion
This proposal allows changes in structured ConfigMap data in JSON format.
```

A little bit more specific and scoped.

Author: I add [story3 and story4](https://github.com/kubernetes-sigs/kustomize/pull/4558/files#diff-2b5a81ecf356afb37e2653ac1e2daf0b6af6cd126aec835df3c25b2a8be29389R342-R447) on my idea.
So, Do you think about needs to be more simple and scoped to this proposal?

## Motivation

<!--
If this proposal is an expansion of an existing GitHub issue, link to it here.
-->

> kustomize can strong patch to Kubernetes objects (yaml file in most cases) with structured edit. And, It is structured data.\

Reviewer: ```suggestion
kustomize can apply structured edits to Kubernetes objects defined in yaml files.
```

> Sometimes structured multi-line or long single line string (ex. json,yaml, and other structured format data) is injected in Kubernetes objects' string literal field. And, kustomize seems it only string literal.\

Reviewer: ```suggestion
Sometimes structured multi-line or long single line string (ex. json,yaml, and other structured format data) is injected in Kubernetes objects' string literal field. From the kustomize perspective, these "structured" multiline strings are just arbitrary unstructured strings.
```

Author: I fix it.

> So, kustomize can't manipulate one value on structured, formatted data in the Kubernetes object's string literal field. This function is expected behavior, but kustomize will be very helpful if it can change the value of structured data like json and yaml substrings in a string literal.\

Reviewer: So kustomize doesn't allow it to be manipulated. This is expected behavior, but we would like to make an exception for some structured JSON within string literal fields.

> Eventually, kustomize doesn't support unstructured edits for string literal. The proposal allows editing of an unstructured string literal as an exception if the string literal is a structured string of a well-known format like json.\

Reviewer: ```suggestion
While kustomize won't support unstructured edits in general, theis proposal allows editing of an unstructured string literal as an exception if the string literal is a structured string of a well-known format like json.
```

Author: I fix it.

This Proposal is add option for [configMapGenerator](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/configmapgenerator/) to allow merge two string literals when the behavior option is setting to merge and string literals value is structured.\
This idea is add one parameter for `valueStructuredMergeFormat` to `option`. The `valueStructuredMergeFormat` option is used by select to structured data format like "json" or "yaml". And this function needs to work requires setting `behavior: merge`.\
This merge operation will be implemented for a part of [Overriding Base ConfigMap Values](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/configmapgenerator/#overriding-base-configmap-values). It will execute to merge two string literal having same [key](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#define-the-key-to-use-when-creating-a-configmap-from-a-file) name when merging two configMap.

#### Example

```yaml
configMapGenerator:
- name: demo-settings
  behavior: merge                     # This function requires `behavior: merge`.
  option:

>     valueStructuredMergeFormat: json  # Setting structured data format.

Reviewer: I see you've made a modification to the original proposal in https://github.com/kubernetes-sigs/kustomize/issues/680#issuecomment-458834785, which looks like:

```
configMapGenerator:
- name: demo-settings
   behavior: merge
   patchContent: json-strategic-merge
- file: my-app-settings=patch.json
- literals:
   - foo= -|
              {
                      "logLevel": "info"
              }
```

I feel that the original proposal is a bit more descriptive, but yours feels a bit more closely tied with the replacements syntax. WDYT about doing this:

```yaml
configMapGenerator:
- name: demo-settings
  behavior: merge                     # This function requires `behavior: merge`.
  option:
    patchContent: strategic-merge
    format: json  # Setting structured data format.
  literals:
  - config.json: |-
      {
        "config": {
          "hostname": "REPLACE_TARGET_HOSTNAME",
          "value": {
            "foo": "bar"
          }
        }
      }
```
?

@KnVerey would also appreciate your feedback on the right syntax here

Author: @natasha41575 

hmm...
I was thinking this implementation will only be used to simply merge 2 or more json contents.
And if we're implementing [json6902](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/patches/#patch-using-inline-json6902) , we need extra sintax like [patchesjson6902](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/patchesjson6902/). 

So, if you have other ideas, could you please tell me?

Author: I can't think this syntax is easy to understand what happen.

```yaml
# overlay/kustomization.yaml
resources:
- ../base
configMapGenerator:
- name: demo
  option:
    valueMergeFormat: json
  literals:
  - config.json: |-
      - op: replace
        path: /config/value/foo
        value: qux
```

Reviewer 2: Sorry for the extreme delay. First, I don't think we should support json6902 or anything other than strategic merge, i.e. the behavior already implied by `behavior: merge` for the rest of the object. That seems sufficient to cover the use case.

Second, I think this option needs to move up a level, i.e. in `GeneratorArgs` rather than `GeneratorOptions`. The latter can be provided globally in a kustomization, and the new data here is specific to the generated object in question.

Third and most importantly, I think we need more information from the user than what the proposals so far have provided for. During overlay configmap generation, we essentially end up with a configmap from the base + a new one from the overlay's files/envs/literals. Currently, `behavior: merge` applies the overlay to the base as an SMP. To support this feature, it seems to me will first need to merge specific fields' values as SMPs before merging the whole object. How do we know which fields we need to do that for? We can't assume they're **all** JSON whenever the user wants one of them to be merged.

```yaml
configMapGenerator:
- name: demo-settings
  behavior: merge
  mergeValues:  # or something like mergeOptions.yamlValues? Not under "options" at any rate
  - config.json # Value at this key MUST be YAML/JSON
  literals:
  - config.json: |-
      {
        "config": {
          "hostname": "REPLACE_TARGET_HOSTNAME",
          "value": {
            "foo": "bar"
          }
        }
      }
```

Finally, a bit of a sidenote: if you try to specify duplicate keys across literal, env and file sources, Kustomize currently throws an error. That behaviour should be retained. In other words, the new feature should only affect the final merge operation.

Author: @KnVerey 
I fix from your comments. At https://github.com/kubernetes-sigs/kustomize/pull/4558/commits/d99011fef31d8f4a1eed8bdfd299a7a9f000cece 
Cloud you recheck it to reflect on your suggestion?

> Second, I think this option needs to move up a level, i.e. in GeneratorArgs rather than GeneratorOptions.

I move it.

> Third and most importantly, How do we know which fields we need to do that for?

I think it is my insufficient consideration and agree with you. I add a parameter to choose a key for the configMap.
But, I think we need any parameter to select a value format which is JSON or YAML. because we can't discriminant clearly the value format when it is not selected with the file extension. For example, value is used by env in deployments like JSON in the environment values.
Therefore I add a parameter `mergeValues.format` to select the value format which is JSON or YAML.
https://github.com/kubernetes-sigs/kustomize/pull/4558/commits/d99011fef31d8f4a1eed8bdfd299a7a9f000cece#diff-2b5a81ecf356afb37e2653ac1e2daf0b6af6cd126aec835df3c25b2a8be29389R153-R155

While kustomize won't support unstructured edits in general, theis proposal allows editing of an unstructured string literal as an exception if the string literal is a structured string of a well-known format like json.\

For example, kustomize can't change the value `"REPLACE_TARGET_HOSTNAME"` in this yaml file straightforwardly.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: target-configmap
data:
  config.json: |-
    {"config": {
      "id": "42",
      "hostname": "REPLACE_TARGET_HOSTNAME"
    }}
```

This function has been wanted for a long time.
- https://github.com/kubernetes-sigs/kustomize/issues/680
- https://github.com/kubernetes-sigs/kustomize/issues/3787
- https://github.com/kubernetes-sigs/kustomize/issues/4517

This function can be an alternative deprecated [vars](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/vars/) function in many use cases.

**Goals:**
<!--
List the specific goals of the proposal. What is it trying to achieve? How will we
know that this has succeeded?
-->

1. Provide the way to update values in the structured data like kubernetes objects with many formats.
1. Be able to add replacement able format after.

> **Non-goals:**

Reviewer 2: Also, we should scope out customizing the field merge strategies for the embedded data, and we should add a brief section to the proposal outlining what that means. For example, lists are always treated as atomic.

> **Non-goals:**

Author: > customizing the field merge strategies for the embedded data

I'm feeling too sorry, I can't understand this mean clearly.
Could you give me additional info like examples?

> **Non-goals:**

Reviewer 3: I'm only guessing, but maybe @KnVerey wants to scope out how the inline strings are merged from this KEP. 

For `behavior: merge` on `configMapGenerator`, this feature will merge the 2 inline, say yaml, strings as kustomize has always merged 2 yaml configurations, instead of allowing the user to specify the merge strategy. In such a case, 

> lists are always treated as atomic

for example.

This idea is add two parameter `format` and `formatPath` to [options](https://github.com/kubernetes-sigs/kustomize/blob/8668691ade05bc17b3c6f44bcd4723735033196e/api/types/replacement.go#L67-L80) in replacement [TargetSelector](https://github.com/kubernetes-sigs/kustomize/blob/8668691ade05bc17b3c6f44bcd4723735033196e/api/types/replacement.go#L52-L64). The `format` option is used by select to structured data format like "json" or "yaml", and The `formatPath` option is "path" to target to change values in structured data with selected format from `format` option.\
I think these two parameters can't select a specific default value. Therefore kustomize return error message for the user if only one parameter was set.\

#### Example.

```yaml
## replacement
replacements:
- source:
    kind: ConfigMap
    name: source-configmap
    fieldPath: data.HOSTNAME
  targets:
  - select:
      kind: ConfigMap
      name: target-configmap
    fieldPaths:
    - data.config\.json
    options:
      format: 'json'                  # Setting structured data format.

>       formatPath: '/config/hostname'  # Setting replacements path.

Author: Thank you!

>       formatPath: '/config/hostname'  # Setting replacements path.

Author: > Could we even make it happen automagically with no additional fields?

My idea is to add an additional syntax to select a value format which is JSON or YAML on `fieldPath` field.
`fieldPath : data.[json:....]`

### example
```yaml
fieldPaths:
- data.[json:[config.json].config.hostname]
```

Do you have any worries about this syntax?

>       formatPath: '/config/hostname'  # Setting replacements path.

Author: > Will create be supported? (yes?)
> Will this be supported in sources as well? (yes?)

I think we need to support it and not dependent on what interface is decided.

> How do these new options interact with the existing options for targeting below the field level, i.e. delimeter and index? (delimiter and index applied last?)

If we select an interface like an extension to fieldPaths, This function is expected to be effective before other `options` and other `options` effect after this. (I think delimiter and index applied last)

>       formatPath: '/config/hostname'  # Setting replacements path.

Reviewer: > How do these new options interact with the existing options for targeting below the field level, i.e. delimeter and index? (delimiter and index applied last?)
> Will create be supported? (yes?)

Agree with @koba1t that we need to support create and index/delimiter options, but that it is independent of the interface. And likewise index and delimiter can be applied last.

> formatPath is effectively an extension to fieldPaths. As such, it should use the same (jsonpath) syntax IMO.

I agree with this one, let's update the syntax of `formatPath` to be the same as the syntax of `fieldPaths`.

> My idea is to add an additional syntax to select a value format which is JSON or YAML on fieldPath field.
fieldPath : data.[json:....]

I do have a concern about the readability of this syntax. It is already kind of hard in a complex yaml file to get the fieldpath syntax correct, and adding this additional syntax to `fieldPath` might make it even harder to use and read.
If we go this route to keep it on `fieldpath`, I prefer @KnVerey's suggestion above:

> For example, they give us fieldPath: data.[config.json].config.hostname, we drill to data.[config.json] and discover it is a string yet we have more path segments remaining, so we try to parse it as yaml. If parsing succeeds, we continue to drill down.

However, I don't think it is _necessary_ to combine this info fieldPath. Another option we can consider is instead of putting it under `options`, we add a new field at the same level at fieldPath and call it `subPath`. If we are thinking to restrict the feature to JSON and YAML (which @KnVerey says above that we should, and I agree), then I don't think we need the `format` field. Maybe something like:

Examples:

Suggestion 1
```
## replacement
replacements:
- source:
    kind: ConfigMap
    name: source-configmap
    fieldPath: data.values.config.hostname # we parse down to data.values and discover a string, but we see that it is JSON formatted so we keep going
  targets:
  - select:
      kind: ConfigMap
      name: target-configmap
    fieldPaths:
    - data.config\.json.config.hostname # we parse down to data.config\.json and discover string, but we see that it is JSON formatted so we keep going
    subPaths:
    - 'config.hostname'
```

Suggestion 2
```
## replacement
replacements:
- source:
    kind: ConfigMap
    name: source-configmap
    fieldPath: data.values
    subpath: 'config.hostname'
  targets:
  - select:
      kind: ConfigMap
      name: target-configmap
    fieldPaths:
    - data.config\.json
    subPaths:
    - 'config.hostname'
```

Do one of these alternatives seem ok for you?

>       formatPath: '/config/hostname'  # Setting replacements path.

Author: Thanks!
I understand your concern about complex syntax.
I want to go with @KnVerey's suggestion due to `subPaths` requiring interface changing, and I think any user wants to manipulate yaml without care about the wall for yaml in yaml.

```

Please check [Story 1](#Story-1).

> ### Disciplined merge the value in structured data with configMapGenerator

Author: I'll add support for secretGenerator.

    fieldPath: data.HOSTNAME
  targets:
  - select:
      kind: ConfigMap
      name: target-configmap
    fieldPaths:
    - data.config\.json
    options:
      format: 'json'
      formatPath: '/config/hostname'
```

```yaml
## expected
apiVersion: v1
kind: ConfigMap
metadata:
  name: source-configmap
data:
  HOSTNAME: www.example.com
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: target-configmap
data:
  config.json: '{"config":{"hostname":"www.example.com","id":"42"}}'
```

#### Story 2

Scenario summary: Merge the two values formatted by structured data(json) with configMapGenerator.

<!--
A walkthrough of what it will look like for a user to take advantage of the new feature.
Include the the steps the user will take and samples of the commands they'll run
and config they'll use.
-->

Many application needs to be set with json format file. And, That file is handled with configMap when the application is running on kubernetes.\
So, If kustomize configMapGenerator can overlay to one line inside a configMap data value, A json format file will be simple and easy to handle.

#### Source
```yaml
# base/kustomization.yaml
configMapGenerator:
- name: demo

>   behavior: merge

Reviewer 2: nit: don't set the merge option in the base

Author: Sorry, I move it in the overlay.

```

```yaml
## replacement
replacements:
- source:
    kind: ConfigMap
    name: environment-config
    fieldPath: data.env
  targets:
  - select:
      kind: ConfigMap
      name: prometheus-config
    fieldPaths:
    - prometheus\.yml
    options:
      format: 'yaml'
      formatPath: '/global/external_labels/prometheus_env'
```

```yaml
## expected
apiVersion: v1
kind: ConfigMap
metadata:
  name: environment-config
data:
  env: dev
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: prometheus-config
data:
  prometheus.yml: |-
    global:
      external_labels:
        prometheus_env: dev
    scrape_configs:
      - job_name: "prometheus"
        static_configs:
          - targets: ["localhost:9090"]
```

#### Story 4

Scenario summary: Replacement the value inside for json in the annotations.
<!--
A walkthrough of what it will look like for a user to take advantage of the new feature.
Include the the steps the user will take and samples of the commands they'll run
and config they'll use.
-->

A few times, an application on your cluster requires to set json format config for the *Annotations* on kubernetes yaml resources. We need to overlay in this position value.

> ##### Example

Reviewer 2: This example seems to be missing the Kustomization and overlay portions

Author: I add it.

https://github.com/kubernetes-sigs/kustomize/pull/4558/commits/83c7492d569dabf15050cfdba254f2717c0ac332#diff-2b5a81ecf356afb37e2653ac1e2daf0b6af6cd126aec835df3c25b2a8be29389R437-R487

This proposal allows the user to identify strings literals containing JSON/YAML data so that Kustomize can make structured edits to the data they contain. This allows the requested functionality to be added without violating Kustomize's core principle of supporting structured edits exclusively.\

For example, kustomize can't change the value `"REPLACE_TARGET_HOSTNAME"` in this yaml file straightforwardly.

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: target-configmap
data:
  config.json: |-
    {"config": {
      "id": "42",
      "hostname": "REPLACE_TARGET_HOSTNAME"
    }}
```

This function has been wanted for a long time.
- https://github.com/kubernetes-sigs/kustomize/issues/680
- https://github.com/kubernetes-sigs/kustomize/issues/3787
- https://github.com/kubernetes-sigs/kustomize/issues/4517

This function can be an alternative deprecated [vars](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/vars/) function in many use cases.

**Goals:**
<!--
List the specific goals of the proposal. What is it trying to achieve? How will we
know that this has succeeded?
-->

> 1. Provide the way to update values in the structured data like kubernetes objects with JSON/YAML formats.

Reviewer 4: I have been observing this for some time because I was bitten by the problem this proposal is trying to address. Unfortunately, I believe, this cannot be solved 100% to the user's expectation.

Various operators put various data in their config maps. My biggest question to this proposal is, what is going to happen when a config map has something called JSON but that JSON is actually a template?

```yaml
# base/kustomization.yaml
configMapGenerator:
- name: demo
  literals:
  - config.json: |-
      {
        "config": {
          "loglevel": debug,
          "parameter": {
            "foo": {{ I am some golang template }}
          }
        }
      }
```

Reviewer 5: Is this proposed idea deprecated by or duplicative of `transformers:` plugins? Maybe more clear that all "opaque" data handling is done in plugins, and not possibly inline? There could be a plugin per transform type like jsonStrategicMerge, json6902, [jq](https://stedolan.github.io/jq/), [yq](https://stedolan.github.io/jq/), etc...

I have templates in config/secret data as well like mentioned by @radekg. The base config uses [mustache](https://mustache.github.io/), and in an overlay I'm applying a custom transformer following the example at https://github.com/badjware/kustomize-plugins/tree/61701852ee3e9873ca9ceed68c0936c28c06d7d5/badjware/v1/placeholdertransformer which pulls in data externally based on my plugin's configuration (from Google Secret Manager). Although I do have to specify `--enable-alpha-plugins` and `KUSTOMIZE_PLUGIN_HOME`, it seems to work great.

Reviewer 6: @radekg I think you ask an important question.

In my principle uses for k-vars today, the base is generated from helm and I have injected uniquely identifiable marker strings into these rendered manifests specifically so that they can be replaced in kustomized overlays using vars.

Replacing strings that have been carefully injected into strangely formatted data is not an edge case.

I have provided the "structure" that is needed for replacement by carefully choosing these marker strings.
